File: markov.hs
import qualified Data.Map as Map
import qualified Data.List as List
import Random (randomRIO)
import Control.Monad (liftM)

------------------------------------------------------------
---- utility functions

-- add a carriage return to each string
--
-- ["a", "b"] => ["a\n", "b\n"]
addCarriageReturns :: [String] -> [String]
addCarriageReturns = map (++ "\n")

-- groupInto 2 "hello" => ["he","el","ll","lo"]
groupInto :: Int -> [a] -> [[a]]
groupInto len = takeWhile isLen . groups
  where
    isLen = (len ==) . length
    groups ls = take len ls : groups (tail ls)

-- takeLast 2 "test" => "st"
takeLast :: Int -> [a] -> [a]
takeLast n ls = drop (length ls - n) ls

-- butLast "test" => "tes"
butLast :: [a] -> [a]
butLast ls = take (length ls - 1) ls

-- return a random element from the list
--
-- "abcde" => "d"
rand :: [a] -> IO a
rand xs = liftM (xs !!) (randomRIO (0, length xs - 1))

------------------------------------------------------------
---- app functions

sampleNames :: [String]
sampleNames = ["Aeneas", "Amadeus", "Andreas", "Antonius", "Apollos", "Atticus", "Augustus", "Aurelius", "Caesar", "Caius", "Cassius", "Cato", "Cicero", "Claudius", "Cornelius", "Cosmo", "Cyrus", "Decimus", "Demetrius", "Felix", "Flavius", "Gaius", "Horatio", "Justus", "Lazarus", "Lucius", "Magnus", "Marcellus", "Marcus", "Marius", "Maximus", "Nero", "Octavius", "Philo", "Primus", "Quintus", "Remus", "Romanus", "Romulus", "Rufus", "Seneca", "Septimus", "Severus", "Stephanus", "Tarquin", "Theon", "Tiberius", "Titus", "Urban"]

type Table = Map.Map String String  -- the markov table
type Order = Int

-- pairize 2 "hello" => [("he","l"), ("el","l"), ("ll","o")]
pairize :: Order -> String -> [(String, String)]
pairize order = map (splitAt order) . groupInto (order + 1)

-- train table with the given left/right pair
train :: (String, String) -> Table -> Table
train (left, right) = Map.insertWith (++) left right

-- train the table with all pairs in the word
trainWord :: Order -> String -> Table -> Table
trainWord order word table = foldr train table (pairize order word)

-- train table with all words
trainWords :: Order -> [String] -> Table
trainWords order = foldr (trainWord order) Map.empty

-- return keys from table that are starting rows
starters :: Table -> [String]
starters = filter startsWithCapital . Map.keys
  where startsWithCapital = flip elem ['A'..'Z'] . head

-- return true if the word has ended
hasEnded :: String -> Bool
hasEnded = (== '\n') . last

-- get markov order of table
getOrder :: Table -> Int
getOrder = length . head . Map.keys

-- parse raw data from file into list of name: ["Tom", "Harry"]
parseFile :: String -> [String]
parseFile = addCarriageReturns . words

-- clean and print a completed word
printWord :: String -> IO ()
printWord = putStrLn . butLast

-- construct a new word, based on the table
buildWord :: Table -> IO String
buildWord table = rand (starters table) >>= buildWord'
  where
    order = getOrder table
    ending = takeLast order
    buildWord' word
      | hasEnded word = return word
      | otherwise = rand options >>= buildWord' . addLetter
      where
        options = table Map.! ending word
        addLetter = (word ++) . (: [])

-- construct an infinite list of words of original words from a table.  The
-- input list of words should well-formatted: they should start with capitals
-- and end with carriage returns.  All words returned will pass the filter
-- function.
buildWords :: (String -> Bool) -> Table -> [IO String]
buildWords filterFn table =
  (buildWord table >>= check) : buildWords filterFn table
    where check word
            | filterFn word = return word
            | otherwise = head (buildWords filterFn table)

-- create a function that determines whether a given word is old
isOld :: [String] -> (String -> Bool)
isOld oldNames = not . flip elem oldNames

main = do
  let order = 2
  let len = 20
  let filename = "male_names.txt"
  rawFileData <- readFile filename
  let oldNames = parseFile rawFileData
  let table = trainWords order oldNames
  let filterFn = isOld oldNames
  newNames <- (sequence . take len . buildWords filterFn) table
  mapM_ printWord newNames

File: markov.clj
;;;; definitions
;;
;; table   : the probability distribution table
;; left    : left side of table, seed; is order tokens long
;; right   : right side of table, all options of what to grow to
;; pair    : a left and right to be trained
;; order   : the order of the markov chain; how far back to consider
;;
;; token   : minimal unit of growth
;; name  : a list of tokens from source material; has beginning and end
;; names : more than one name, i.e., more than one source

(def test-order 2)
(def test-names (map #(str % "\n") [ "Aeneas" "Amadeus" "Andreas" "Antonius" "Apollos" "Atticus" "Augustus" "Aurelius" "Caesar" "Caius" "Cassius" "Cato" "Cicero" "Claudius" "Cornelius" "Cosmo" "Cyrus" "Decimus" "Demetrius" "Felix" "Flavius" "Gaius" "Horatio" "Justus" "Lazarus" "Lucius" "Magnus" "Marcellus" "Marcus" "Marius" "Maximus" "Nero" "Octavius" "Philo" "Primus" "Quintus" "Remus" "Romanus" "Romulus" "Rufus" "Seneca" "Septimus" "Severus" "Stephanus" "Tarquin" "Theon" "Tiberius" "Titus" "Urban" ]))
(def test-start-chars "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
(def test-end-chars ["\n"])

(defn train
  "Take a table and a pair, and return a table trained with that pair."
  [table [left right]]
  (update-in table [left] conj right))

(defn pairize
  "Take a name and an order, and return a list of all possible pairs."
  [name order]
  (for [chars (partition (inc order) 1 name)]
    [(butlast chars) (last chars)]))

(defn train-names
  "Create a probability distribution table based on list of names."
  [names order]
  (reduce (fn [table name]
            (reduce train table (pairize name order)))
          {}
          names))

(defn in?
  "Is n in coll?"
  [n coll] (some (partial = n) coll))

(defn starters
  "Return all starting pairs in table."
  [table start-chars]
  (for [[left right :as pair] table
        :when (in? (first left) start-chars)]
    left))

(defn done?
  "Return if name is complete."
  [name end-chars]
  (not (in? (last name) "abcdefghijklmnopqrstuvwxyz")))

; return random element of list
(def rnd (comp first shuffle))

; (load-file "markov.clj")

(defn build [table order start-chars end-chars]
  (let [start-rows (starters table start-chars)]
    (loop [word (apply str (rnd start-rows))]
      (let [letter (rnd (get table (take-last order word)))
            new-word (str word letter)]
        (if (done? new-word end-chars)
          new-word
          (recur new-word))))))

(def male-names (map #(str % "\n") (clojure.string/split (slurp "male_names.txt") #"\n")))

(dotimes [n 20]
  (println (apply str (butlast (build (train-names male-names 2)
       2 test-start-chars test-end-chars)))))

